WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/data/models/WooPosProductModelVersion2.kt

```
@@ -0,0 +1,118 @@
+package com.woocommerce.android.ui.woopos.common.data.models
+
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+import java.math.BigDecimal
+
+/**
+ * This model provides a clean separation between the data layer (WCPosProductEntity)
+ * and the view layer, ensuring all required data is present and properly typed.
+ */
+@Parcelize
+data class WooPosProductModelVersion2(
+    val id: Long,
+    val parentId: Long?,
+    val name: String,
+    val sku: String,
+    val globalUniqueId: String,
+    val type: WooPosProductType,
+    val status: WooPosProductStatus,
+    val pricing: WooPosPricing,
+    val description: String,
+    val shortDescription: String,
+    val isDownloadable: Boolean,
+    val lastModified: String,
+    val images: List<WooPosProductImage> = emptyList(),
```

🤔 I have been wondering whether these attributes should by parsed lazily - however, we'd have to have access to a gson instance. Passing it in feels off and creating it per model feels of as well. Which leaves me with a single instantiation within companion object, wdyt?

---

I know we already store attributes in `WCPosProductEntity` table, but do we really need them in POS? Can we remove them, maybe, instead of passing them down to the domain layer?

---

Good question. I have been following pdfdoF-7VS-p2 and plan to remove unnecessary columns in https://linear.app/a8c/issue/WOOMOB-1120/woo-poslocal-catalog-remove-unnecessary-columns-from-pos-products. It's difficult to say now which columns will be needed and which not. What do you think about this approach?

---

But yeah, I agree it would remove the need to think about lazy loading now. Let me add a note to that ticket to also consider the lazy loading when we know more.

---

Anyway, I'm not a fan of lazy parsing, and would propose parsing JSON while mapping the entity. This would give us:
* Type safety and clean models
* Fail fast advantage: JSON parsing errors caught early during mapping, not during UI rendering
* Performance?: I'm not 100% sure, but maybe in some cases, lazy parsing would cause worse performance when the same data is accessed multiple times.

Wdyt?

---

> Performance?: I'm not 100% sure, but maybe in some cases, lazy parsing would cause worse performance when the same data is accessed multiple times.

Lazy parsing never parses it twice - it always parses it at most once, but only when it's needed and caches the result => so lazy parsing definitely provides a better performance. 

> Fail fast advantage: JSON parsing errors caught early during mapping, not during UI rendering

Having said that, I agree with the fail-fast advantage. Even though we might just want to consume parsing errors in these attributes anyway. But still feels slightly better. Let's keep this current approach and migrate to lazy if we notice performance issues.